File: Proyecto1/Proyecto1/coman_mkdisk.cpp
#include "coman_mkdisk.h"
#include <stdio.h>
#include <string.h>
#include <iostream>
#include <time.h>
#include "string"
#include <QString>
using namespace std;

coman_mkdisk::coman_mkdisk()
{

}

void coman_mkdisk::recorrido_mkdisk(Nodo_arbol *raiz){
    bool bandera_size = false;
    bool bandera_path = false;
    bool bandera_error = false;
    int valor_size = 0;
    string valor_fit = "ff";
    string valor_unit = "m";
    QString valor_path = "";
    for(int c = 0; c < raiz->hijos.count();c++){
        Nodo_arbol temp = raiz->hijos.at(c);
        if(temp.tipo_num == SIZE){
            if(temp.valor.toUInt()>0){
                valor_size = temp.valor.toUInt();
                bandera_size = true;
            }else{
                bandera_error = true;
                std::cout<<"Los valores de size no pueden ser menores o iguales a cero \n";
            }
        }
        else if(temp.tipo_num == FIT){
            valor_fit = temp.hijos.at(0).valor.toStdString();
        }
        else if(temp.tipo_num == UNIT){
            valor_unit = temp.hijos.at(0).valor.toStdString();
        }
        if(temp.tipo_num == PATH){
            bandera_path = true;
            valor_path = temp.valor;
        }
    }
    if(bandera_size == false){
        bandera_error = true;
        std::cout<<"Error: El parametro -size es obligatorio \n";
    }
    if(bandera_path == false){
        bandera_error = true;
        std::cout<<"Error: El parametro -path es obligatorio \n";
    }
    if(bandera_error == false){
        ejecutarDisco(valor_size, valor_fit, valor_unit, valor_path);
    }
}

void coman_mkdisk::ejecutarDisco(int psize, string pfit, string punit, QString ppath){

    QString cambio_ruta = ppath;
    cambio_ruta.replace(QString("\""), QString(""));
    string rutax = cambio_ruta.toStdString();

    mbr prueba;
    creandoRuta(cambio_ruta);

    FILE *archivo;
    archivo = fopen(rutax.c_str(), "rb+");

    //Simulando un Kb
    char buffer[1024];

    //Si son Kb
    if(punit == "k" || punit == "K"){
        //Este if es para el tamaño
        prueba.mbr_tamano = psize*1024;

        for(int i = 0; i<1024 ; i++){
            buffer[i] = '\0';
        }

        for(int i = 0; i<(psize); i++){
            fwrite(&buffer, 1024, 1, archivo);
        }
        fclose(archivo);
    }

    //Si son megas
    else if(punit == "m" || punit == "M"){
        prueba.mbr_tamano = psize*1024*1024; //En bytes

        for(int i = 0; i<1024; i++){
            buffer[i] = '\0';
        }
        for(int i = 0; i<(psize*1024); i++){
            fwrite(&buffer, 1024, 1, archivo);
        }
        fclose(archivo);
    }

    prueba.mbr_fecha_creacion = time(nullptr);
    prueba.mbr_dsk_signature = (rand()% 100);//Numero random para la etiqueta del disco

    char auxfit = 0;
    if(pfit == "bf" || "BF"){
        auxfit = 'B';
    }else if(pfit == "ff" || pfit == "FF"){
        auxfit = 'F';
    }else if(pfit == "wf" || pfit == "WF"){
        auxfit = 'W';
    }

    prueba.dsk_fit = auxfit;

    for(int p = 0; p < 4; p++){
        prueba.mbr_partition[p].part_status = '0';
        prueba.mbr_partition[p].part_type = '0';
        prueba.mbr_partition[p].part_fit = '0';
        prueba.mbr_partition[p].part_size = 0;
        prueba.mbr_partition[p].part_start = -1;
        strcpy(prueba.mbr_partition[p].part_name,"");
    }

    archivo = fopen(rutax.c_str(), "rb+");
    fseek(archivo, 0, SEEK_SET);
    fwrite(&prueba, sizeof(mbr), 1, archivo);
    fclose(archivo);
    cout<<"Disco creado correctamente"<<endl;
    cout<<"Se agrego el mbr de manera correcta \n"<<endl;
}

void coman_mkdisk::creandoRuta(QString path){
    QString aux = obtener_path(path);
    string comando = "sudo mkdir -p \'"+aux.toStdString()+"\'";
    cout << comando << endl;
    system(comando.c_str());
    string comando2 = "sudo chmod -R 777 \'"+aux.toStdString()+"\'";
    system(comando2.c_str());
    string arch = path.toStdString();
    FILE *archivo = fopen(arch.c_str(),"wb");
    if((archivo = fopen(arch.c_str(),"wb")))
        fclose(archivo);
    else
        cout << "Error: Al crear el Disco" << endl;
}

QString coman_mkdisk::obtener_path(QString path){

    string aux = path.toStdString();
    string delimiter = "/";
    size_t pos = 0;
    string res = "";
    while((pos = aux.find(delimiter))!=string::npos){
        res += aux.substr(0,pos)+"/";
        aux.erase(0,pos + delimiter.length());
    }
    cout << res << endl;
    return QString::fromStdString(res);
}

File: Proyecto2/analizador/comandos/coman_mkdisk.go
package comandos

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
	"unsafe"
)

func Recorrido_mkdisk(comando string) {
	sinsalto := strings.TrimRight(comando, "\n")
	coman := strings.Split(sinsalto, " ")
	var bandera_size bool = false
	var bandera_path bool = false
	var bandera_error bool = false
	var valor_size int = 0
	var valor_fit string = "ff"
	var valor_unit string = "m"
	var valor_path string = ""
	for i := 0; i < len(coman); i++ {
		param := strings.Split(coman[i], "=")
		for j := 0; j < len(param); j++ {
			if param[j] == "-size" || param[j] == "-SIZE" {
				sv, _ := strconv.Atoi(param[j+1])
				valor_size = sv
				if valor_size <= 0 {
					bandera_error = true
					fmt.Println("Los valores de size no pueden ser menores o iguales a cero")
				} else {
					bandera_size = true
				}
			} else if param[j] == "-fit" || param[j] == "-FIT" {
				valor_fit = param[j+1]
			} else if param[j] == "-unit" || param[j] == "-UNIT" {
				valor_unit = param[j+1]
			} else if param[j] == "-path" || param[j] == "-PATH" {
				bandera_path = true
				valor_path = param[j+1]
			}
		}
	}
	if !bandera_size {
		bandera_error = true
		fmt.Println("Error: El parametro -size es obligatorio")
	}
	if !bandera_path {
		bandera_error = true
		fmt.Println("Error: El parametro -path es obligatorio")
	}
	if !bandera_error {
		ejecutarDisco(valor_size, valor_fit, valor_unit, valor_path)
	}
}

func ejecutarDisco(psize int, pfit string, punit string, ppath string) {

	comillaDer := strings.TrimRight(ppath, "\"")
	comillaIzq := strings.TrimLeft(comillaDer, "\"")
	cambio_ruta := comillaIzq

	crearDirectorio(cambio_ruta)

	file, _ := os.Create(cambio_ruta)
	defer file.Close()

	prueba := MBR{}

	var temporal int8 = 0
	s := &temporal
	var binario bytes.Buffer
	binary.Write(&binario, binary.BigEndian, s)

	if punit == "k" || punit == "K" {
		prueba.Mbr_tamano = int64(psize * 1024)

		for i := 0; i < psize*1024; i++ {
			escribirBytes(file, binario.Bytes())
		}
	} else if punit == "m" || punit == "M" {
		prueba.Mbr_tamano = int64(psize * 1024 * 1024)

		for i := 0; i < psize*1024*1024; i++ {
			escribirBytes(file, binario.Bytes())
		}
	}

	tiempo := time.Now()
	time := tiempo.String()

	for t := 0; t < 19; t++ {
		prueba.Mbr_fecha_creacion[t] = time[t]
	}

	numeroRandom := rand.Intn(100-1) + 1
	prueba.Mbr_dsk_signature = int64(numeroRandom)

	var auxfit byte = 0

	if pfit == "bf" || pfit == "BF" {
		auxfit = 'B'
	} else if pfit == "ff" || pfit == "FF" {
		auxfit = 'F'
	} else if pfit == "wf" || pfit == "WF" {
		auxfit = 'W'
	}

	prueba.Dsk_fit = auxfit

	for p := 0; p < 4; p++ {
		prueba.Mbr_partition[p].Part_status = '0'
		prueba.Mbr_partition[p].Part_type = '0'
		prueba.Mbr_partition[p].Part_fit = '0'
		prueba.Mbr_partition[p].Part_size = 0
		prueba.Mbr_partition[p].Part_start = -1
		for n := 0; n < 16; n++ {
			prueba.Mbr_partition[p].Part_name[n] = '0'
		}
	}

	//nos posicionamos al inicio del archivo usando la funcion Seek
	file.Seek(0, 0)

	//Escribimos struct de mbr
	var bufferControl bytes.Buffer
	binary.Write(&bufferControl, binary.BigEndian, &prueba)
	escribirBytes(file, bufferControl.Bytes())

	//movemos el puntero a donde ira nuestra primera estructura
	file.Seek(int64(unsafe.Sizeof(prueba)), 0)

	fmt.Println("Disco creado correctamente")
	fmt.Println("Se agrego el mbr de manera correcta")

}

func crearDirectorio(path string) {
	directorio := obtener_path(path)
	if _, err := os.Stat(directorio); os.IsNotExist(err) {
		err = os.MkdirAll(directorio, 0777)
		if err != nil {
			panic(err)
		}
	}
}

func obtener_path(path string) string {
	var aux_path int
	var aux_ruta string

	for i := len(path) - 1; i >= 0; i-- {
		aux_path++
		if string(path[i]) == "/" {
			break
		}
	}

	for i := 0; i < ((len(path)) - (aux_path - 1)); i++ {
		aux_ruta += string(path[i])
	}

	return aux_ruta
}

func escribirBytes(file *os.File, bytes []byte) {
	_, err := file.Write(bytes)

	if err != nil {
		log.Fatal(err)
	}
}
